README
1. General Information:
---------------------------------
   Use your World Wide Web browser to read:

	http://opencircuitdesign.com/magic/
	http://vlsi.csl.cornell.edu/magic/
	http://www.research.digital.com/wrl/magic/magic.html

   Primary documentation is on the opencircuitdesign.com website under
   the "Documentation" link.

   The current development versions of magic are maintained by Tim Edwards
   <tim@opencircuitdesign.com> and the current distribution version is
   maintained by Rajit Manohar <rajit@csl.cornell.edu>.  Please let us
   know of any problems/bugs you find.  Development of versions 7.2 and
   newer is generously funded by MultiGiG, Inc.

2. Compilation and Installation:
---------------------------------
   See the file "INSTALL" in this directory.

3. Version 8.1 Release Notes:
---------------------------------

   As of the release of version 8.1, Version 8.0 is now the new stable
   distribution, and 8.1 is the development distribution.

   What's new in 8.1:
   ------------------
   1) Substantially revised substrate handling allows for "soft
      connectivity" detection, and resolves (finally) the problem of
      the hack handling of nMOS bulk connections using a default
      name without any understanding of a substrate node and
      connectivity.

4. Version 8.0 Release Notes:
---------------------------------

   As of the release of version 8.0, Version 7.5 is now the new stable
   distribution, and 8.0 is the development distribution.

   What's new in 8.0:
   ------------------
   1) Vector outline fonts with ability to control font, scale, rotation,
      and offset.  Public-license outline fonts provided by the freefont
      project.

   2) Hierarchical SPICE output for LVS.

   3) New cifoutput operators including "net" and "maxrect", specifically
      for using with the "cif paint" command.

   4) DRC method for handling via rules specifying overlap on two
      opposite sides but not on the others.

   5) Improved DRC method that ignores errors in subcells that are
      masked by the parent cell.

   6) Improved cell manager based on methods available in Tk 8.5

   7) New extraction method "msubcircuit" with methods for specifying
      parameter names for source/drain area and perimeter.

5. Version 7.5 Release Notes:
---------------------------------

   Version 7.5 is the development branch.  Version 7.5.0 is the same as
   7.4.2, which is essentially the same as 7.3.123, plus some
   documentation updates.  Intended development is as follows:

   What's new in 7.5:
   ------------------
   1) Use a finely spaced grid for the database, but keep the concept
      of "lambda" for layout.  Keep backwards compatibility, and resolve
      issues with layout that does not work well on the lambda grid.
      Implemented in 7.5.1 by allowing a DRC "scalefactor" line,
      which declares that all DRC rules are in units of (lambda /
      scalefactor).  Rules "in use" are scaled to lambda and rounded to
      the nearest integer.  The original value is retained, however, so
      that any call to "scalegrid" will recompute the DRC distances based
      on the current internal grid.  Thus, we can define DRC rules in
      fractional lambda and therefore match vendor DRC rule distances
      while still maintaining magic's concept of "lambda".  This means
      that users working entirely within magic have scalable CMOS rules,
      but if a "vendor cell" (3rd party GDS or CIF) is loaded, the DRC
      rules will be correct with respect to it.
   2) Multiple DRC styles allowed in the technology file.
   3) Memory-mapped tile allocation using the mmap() function.
   4) Layer and cell instance locking
   5) Euclidean-distance measure on "cif grow" operator.
   6) "cif paint" command to automatically manipulate the database
      paint using "cifoutput" rulesets.
   7) New contact-cut generation algorithm.
   8) Added the ability to define and extract MOS devices with
      asymmetric source and drain.
   9) Added extraction devices "rsubcircuit" and "subcircuit" to
      produce subcircuit records in SPICE output, with a method to
      define parameters to be passed to the subcircuit.
  10) Added resistor corner scaling (i.e., the resistance of a
      material at a corner can be set as a fraction of the resistance
      of the same material on a straight path).
  11) Updated the interactive maze router, fixing many bugs, and adding
      many enhancements, including a maze router GUI that can be used
      to aid in interactively routing an entire netlist, or performing
      a verification of a netlist against the layout.
  12) "gridlimit" keyword in the cifoutput section to prevent magic
      from generating geometry beyond a specific resolution.
  13) Added the ability to specify all units in the extract section in
      microns, and added a simplified method for specifying standard
      parasitic capacitance extraction rules.
  14) "gds merge true" option to generate polygons in the GDS output
      instead of tiles.  This creates *much* smaller output files at
      the expense of processing time.
  15) New "contact" function to automatically contact two layers at
      an intersection.

   See the online release notes for a more thorough list of features.

6. Version 7.4 Release Notes:
---------------------------------

   Version 7.4 is the new stable distribution version of magic.
   Apart from changes to the release notes, it is identical to
   the last revision (123) of development version 7.3.  Revisions
   of 7.4 will be made as necessary to fix bugs in the code.  All
   new additions and major changes will be done to the new
   development distribution, version 7.5.  Therefore there will
   not be a "What's new in 7.4" section, as there is not supposed
   to be anything new in version 7.4.

7. Version 7.3 Release Notes:
---------------------------------

   Magic release 7.3 incorporates a stacked contact model which is,
   for the most part, backwardly compatible with Magic releases
   7.2 and earlier.  Information about this developmental release
   can be found at:

	http://opencircuitdesign.com/magic/magic7_3.html


   What's new in 7.3:
   ------------------
   Provided by Tim Edwards (MultiGiG, Inc.):
         1) Stacked contact model allowing arbitrary stacking of
	    contact types.
         2) A separate "undo/redo" method for network selection, to
	    remove the memory overhead associated with selecting and
	    unselecting large networks.  Also removes some time overhead
	    as well, especially when unselecting networks.
         3) Much improved "plot pnm" function.
         4) Improved transistor and resistor extraction.
	 5) LEF format reader; improved LEF/DEF input/output handling
	 6) New style and colormap file formats
	 7) Vendor GDS read/write capability
         8) "wire segment" drawing function
	 9) Handling of path records in CIF and GDS input
	10) Handling of cell scaling in GDS input
	11) Pi-network device extraction for resistors
	12) Option to write contacts as cell arrays in GDS output
	13) New "widespacing" and "maxwidth" DRC algorithms.
	14) "polygon" command
	15) New cifoutput operator "bloat-all"
	16) Backing-store for 24-bit and OpenGL graphics
	17) New "pick" tool for interactive selection move and copy
	18) New interactive "wire" tool
	19) Crosshair
	20) New cifoutput operator "slots"
	21) New fcntl-based file locking mechanism
	22) "angstroms" units supported in cifinput/cifoutput
	23) Non-Manhattan device extraction support
	24) New "feedback" mechanism
	25) Proper support for > 32 planes (up to 64)
	26) Fixed array interaction CIF/GDS generation
	27) Added executable "magicdnull" for streamlined batch-mode use
	28) New method for crash backups, including restore with "magic -r"
        29) A number of other technology file additions and enhancements

8. Version 7.2 Release Notes:
---------------------------------

   Magic release 7.2 incorporates the capability to run magic from the Tcl
   interpreter with graphics handled by Tk.  Instructions for compiling
   and installing this version are in README.Tcl.  Information about
   this release can be found at:

	http://opencircuitdesign.com/magic/magic7_2.html

   What's new in 7.2:
   ------------------
   Provided by Tim Edwards (MultiGiG, Inc., and JHU Applied Physics Lab):

       1) Tcl interpreter extension option
       2) Cygwin compile option
       3) Memory resources cleaned up
       4) GUI interface to Tcl version of Magic
       5) Readline update to version 4.3
       6) OpenGL fixes and refinements
       7) Nonmanhattan geometry fixes and extensions
       8) Threaded graphics in non-Tcl environments
       9) Inductance extraction
      10) CIF and GDS input/output support for deep submicron technologies
      11) Different internal and lambda grids, including automatic or
	  induced ("scalegrid" command) grid subdivision and expansion.
	  "snap" and "grid" functions and extensions aid layout when
	  lambda and internal units differ.
      12) Removed commands "list", "listall", "parent", and "child",
	  replacing them with the more general-purpose "cellname"
	  and "instance" commands.
      13) Added command "tech", and re-loadable technologies.
      14) Revamped the "dstyle" files and updated the dstyle version
      15) Added "element" types for layout annotation.
      16) Extended extract section of techfile to include "device"
	  keyword and devices "mosfet", "bjt", "capacitor", and "resistor".
	  New model resistor and mosfet use width/length instead of area/
	  perimeter.
      17) Added 3D rendering window invoked by command "specialopen wind3d",
	  for the Tcl version compiled with OpenGL graphics.
      18) Added "height" keyword to tech file for height/thickness values
      19) Added "windowname" command for managing multiple wrapper windows
	  under Tcl.
      20) Added extraction extension for annular (ring) MOSFETs.
      21) Added "widespacing" DRC rule.
      22) Added GNU autoconf compile
      23) New command "property" for setting key:value pair properties
	  in cell definitions that can be interpreted by other routines
	  (such as LEF/DEF).
      24) General-purpose subcircuit method using the "port" command to
	  declare a cell to be a subcircuit and to mark the position and
	  orientation of connections into the subcell.  This replaces a
	  method previously built into version 7.2 using a "subcircuit"
	  layer; that method is now considered obsolete.
      25) LEF and DEF format readers, LEF format writer.
      26) Improved techfile format with asterisk-notation and DRC
	  "surround", "overhang", and "rect_only" statements.

9. Version 7.1 Release Notes:
---------------------------------

   Magic release 7.1 consolidates all known patches/features
   to magic version 6.5.x, and contains additional features that have
   been added since the consolidation.  Information about this release
   is available at the magic web site:

	http://vlsi.cornell.edu/magic/


   What's new in 7.1:
   ------------------
   Provided by Rajit Manohar (Cornell University) (formerly version 7.0):
       1) Implementation of "scheme" (a subset of lisp), a powerful method
	  of generating complex functions.
       2) Using CVS to facilitate source code development from multiple sites.
       3) New commands using scheme:  Too many to mention here;  see the
	  tutorials in doc/tutscm* for explanations.  Functions themselves
	  are defined in ${CAD_ROOT}/magic/scm/*.scm
       4) Overhauled the readline interface. See doc/textfiles/readline.txt for
          details.
       5) Changed tons of stuff about the build environment:
            - the include paths in all files is now canonical.
            - redid the make process and rewrote all Makefiles.
            - tons of other small things that hopefully make the build process
              nicer.

10. Releases prior to version 7:
---------------------------------

   What's new in 6.5.2:
   --------------------
   Provided by R. Timothy Edwards (Johns Hopkins Applied Physics Laboratory):
        1) Support for OpenGL 
           Look at doc/open_gl.txt
        2) Minor update to :config for selection of multiple graphics
           interfaces.
        3) Updates to dstyle and cmap files
        4) Always do a check to see if there is only one active layout
           window:  There's no need to annoy the user with complaints
           of "Put the cursor in a layout window" when there is no  
           possibility of confusion about the matter.
    
   Provided by Philippe Pouliquen (Johns Hopkins University):
        5) "readline" command line editing capability
        6) Macro extensions for X11 (see doc/macro_extension.txt)
        7) Better handling of filenames on the UNIX command-line (multiple
           filenames allowed, ".mag" extension not required).
        8) New commands:  "child", "parent", "down", "xload", "list", "listall",
           "listtop", "shell", "imacro".
        9) Command alterations: "box [+|-][dir]", "select visible", area of box
           included in "box" command.
       10) Updated .magic macro file (source in magic/proto.magic, install in
           ${CAD_ROOT}/magic/sys/.magic) (see doc/default_macros.txt).

   What's new in 6.5.1:
   --------------------
	1) Support for true-color displays (courtesy of Michael Godfrey)
	   Look into doc/hires-color.txt
	2) Minor updates in ext2sim, ext2spice

   What's new in 6.5:
   -----------------
        1) Bug fixes in the extractor propagation of attributes (SU)

        2) New version of ext2sim ext2spice with support for hspice, spice2, 
           and spice3 (SU)

        3) Integration of the port to SUN's Solaris OS (MIT)

        4) Port to FreeBSD2.x. Thanks to John Wehle (john@jwlab.feith.com)

	5) Integration of part of the DEC/WRL code fragments into the drc
	   code. Since the code is not completely trustworthy the fragments 
	   are ifdef'd so if need be the drc will behave exactly as the old one.
	   (you just need to change the #define DRC_EXTENSIONS in drc/drc.h
	    to do that).  For a description of the extensions look into 
	    doc/tutwrl1.ps (DEC/WRL)
	
	6) Integration of some patches in to the CIF code that introduce:
	    (i)  A new cif operation squares-grid which generates contacts 
		 on grid.
            (ii) A new cif layer command min-width is added so that generated
		 layers have the correct min drc width when magic expands
		 layers in the hierarchy (like it does with wells).
		 

   Magic-6.5 crashes if compiled with gcc in Solaris2.x/SunOS5.x (curiously enough 
   it does not have that problem if compiled with Sun's cc compiler). To get 
   around that you need to set the flag -DUSE_SYSTEM_MALLOC flag in misc/DFLAGS 
   after you run make :config. The error has to do with allignment of doubles 
   and an alternative way to get rid of it is to change extract/extractInt.h 
   so that CapValue is float instead of double.  Nevertheless the first method 
   is recomended.

   What's new in 6.4:
   ------------------
   This release, magic 6.4, contains the following modifications:

	1) A number of bug fixes from the 6.3 notes.

	2) A version numbering facility for tech files.  Please add a new 
	   section to each tech file after the "tech" section, following 
	   this example:

		version
		    version 2.0.3
		    description "MOSIS CMOS 0.13u Nano-nano technology."
		end
		    
	   Older versions of magic will complain about the new section, but
	   no harm will be done.

	4) Various comments describing dates and versions, including the
	   above tech file information, are not written to the CIF file.

	3) Support for patches and versioning:  A new command called "version"
	   lists out the version number and patches that are installed.
	   A header file called patchlevel.h keeps track of a PATCHLEVEL 
	   integer and a string in patchlevel.c keeps track of the names of
	   each installed patch.  When posting patches to the net please be
	   sure your patch updates variables.  See the files for details.

	4) Ports to Alpha AXP OSF/1, SGI/IRIX (courtesy of Stefanos 
	   Sidiropoulos) and Linux (courtesy of Harold Levy).

	5) A change in the extractor algorithm to provide shielding for
	   perimeter capacitances. Also a change in ext2sim to maintain
	   information about the area and perimeter of diffusion and
	   the bulk connection of the fets (written by Stefanos
	   Sidiropoulos).


debian/changelog
magic (8.1.214+ds.1-1) unstable; urgency=low

  * New upstream release
  * debian/control:
    - Priority changed from extra to optional
    - Maintainer changed to Debian Science Maintainers (Closes: #812749)
    - Myself added as uploader
    - dh-autoreconf and libglu1-mesa-dev added as build-dependencies
    - Standards version updated to 4.1.5
    - Added new VCS URLs
    - Fix whitespace error
  * debian/copyright:
    - Format field uses https
    - Updated some copyright years.
    - Added myself to listing for debian/*
    - Added Files-Excluded for embedded readline library
  * debian/gbp.conf:
    - Specify that pristine-tar is used
  * debian/lintian-overrides
    - Override some wrongly detected lintian errors
  * debian/patches:
    - 0008-Some-spelling-errors-fixed.patch - fix spelling errors
    - 0009-Tutorials-and-ps-files-installed-in-usr-share-doc-ma.patch
      - Put these files in FHS compliant directories
    - Refreshed all other patches and make them work with gbp-pq
  * debian/README.source:
    - Specify what is excluded in repacked source
  * debian/rules:
    - Set hardening flags
    - Use autoreconf instead of autools_dev
    - Added debian/autoreconf to specify directory to run autoreconf
  * debian/watch:
    - Add dversionmangle and oversionmangle in order to properly add +ds1
      suffix to upstream version
    - Updated watch file so that it looks for all versions in the 8.1.x range.
      Earlier it was watching only 7.x versions.

 -- Ruben Undheim <ruben.undheim@gmail.com>  Sun, 29 Jul 2018 11:40:39 +0200

magic (8.0.210-2) unstable; urgency=medium

  * Added 08-reproducible-build.patch to support reproducible builds,
    thanks to Chris Lamb (Closes: #778272)

 -- Roland Stigge <stigge@antcom.de>  Sun, 17 Jan 2016 14:16:27 +0100

magic (8.0.210-1) unstable; urgency=medium

  * New upstream release
  * debian/control: Standards-Version: 3.9.6

 -- Roland Stigge <stigge@antcom.de>  Sat, 16 Jan 2016 17:20:41 +0100

magic (7.5.241-1) unstable; urgency=medium

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Mon, 13 Oct 2014 11:11:14 +0200

magic (7.5.240-1) unstable; urgency=medium

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Sun, 31 Aug 2014 14:51:35 +0200

magic (7.5.239-1) unstable; urgency=medium

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Sat, 07 Jun 2014 16:16:29 +0200

magic (7.5.236-1) unstable; urgency=medium

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Thu, 29 May 2014 18:28:51 +0200

magic (7.5.234-1) unstable; urgency=medium

  * New upstream release
  * debian/control:
    - Standards-Version: 3.9.5
    - Add wish and tclsh to Depends

 -- Roland Stigge <stigge@antcom.de>  Sun, 18 May 2014 19:38:24 +0200

magic (7.5.233-2) unstable; urgency=low

  * Use autotools-dev to update config.{sub,guess} (Closes: #727452)

 -- Roland Stigge <stigge@antcom.de>  Thu, 24 Oct 2013 11:33:36 +0200

magic (7.5.233-1) unstable; urgency=low

  * New upstream release
  * Use readlink for /usr/bin/wish in magic to use versioned wish

 -- Roland Stigge <stigge@antcom.de>  Thu, 17 Oct 2013 14:36:32 +0200

magic (7.5.231-1) unstable; urgency=low

  * New upstream release
  * debian/control: Standards-Version: 3.9.4

 -- Roland Stigge <stigge@antcom.de>  Wed, 15 May 2013 21:56:15 +0200

magic (7.5.230-1) experimental; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Fri, 25 Jan 2013 11:10:07 +0100

magic (7.5.229-1) experimental; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Mon, 08 Oct 2012 09:35:33 +0200

magic (7.5.228-1) experimental; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Wed, 03 Oct 2012 12:14:31 +0200

magic (7.5.225-1) experimental; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Tue, 25 Sep 2012 13:13:14 +0200

magic (7.5.220-1) unstable; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Wed, 02 May 2012 20:45:37 +0200

magic (7.5.218-1) unstable; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Sat, 21 Apr 2012 09:26:03 +0200

magic (7.5.216-1) unstable; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Fri, 06 Apr 2012 14:47:13 +0200

magic (7.5.215-1) unstable; urgency=low

  * New upstream release
  * debian/control: Standards-Version: 3.9.3
  * debian/compat: 9

 -- Roland Stigge <stigge@antcom.de>  Sat, 10 Mar 2012 20:47:56 +0100

magic (7.5.214-1) unstable; urgency=low

  * New upstream release

 -- Roland Stigge <stigge@antcom.de>  Sun, 27 Nov 2011 22:05:55 +0100

magic (7.5.213-1) unstable; urgency=low

  * Initial release (Closes: #365326)

 -- Roland Stigge <stigge@antcom.de>  Tue, 26 Jul 2011 21:53:03 +0200


cif/CIFint.h
/*
 * CIFint.h --
 *
 * Defines things shared internally by the cif module of Magic,
 * but not generally needed outside the cif module.
 *
 *     ********************************************************************* 
 *     * Copyright (C) 1985, 1990 Regents of the University of California. * 
 *     * Permission to use, copy, modify, and distribute this              * 
 *     * software and its documentation for any purpose and without        * 
 *     * fee is hereby granted, provided that the above copyright          * 
 *     * notice appear in all copies.  The University of California        * 
 *     * makes no representations about the suitability of this            * 
 *     * software for any purpose.  It is provided "as is" without         * 
 *     * express or implied warranty.  Export of this software outside     * 
 *     * of the United States of America may require an export license.    * 
 *     *********************************************************************
 *
 * rcsid "$Header: /usr/cvsroot/magic-7.5/cif/CIFint.h,v 1.5 2008/11/25 22:49:32 tim Exp $"
 */

#ifndef _CIFINT_H
#define _CIFINT_H

#include "database/database.h"

/* The main data structure used in the cif module is a description
 * of how to generate CIF layers from the Magic tiles.  There may
 * be several different styles for generating CIF from the same Magic
 * information, e.g. for fabricating at different geometries.  Each
 * of these CIF styles involves three kinds of data.  A "CIFStyle"
 * record gives overall information such as the number of layers.
 * One "CIFLayer" gives overall information for each layer, and
 * then a list of one or more "CIFOp" records describes a sequence
 * of geometrical operations to perform to generate the layer.  This
 * data structure is built up by reading the technology file.
 */

/* A CIFOp starts from a partially-completed CIF layer, does something
 * to it, which may possibly involve some existing layers or temporary
 * layers, and creates the next stage of the partially-completed
 * CIF layer.  Example operations are to AND with some existing paint,
 * or to grow by a certain amount.
 */

typedef struct bloat_data
{
    int bl_plane;		/* Plane on which a bloat or squares
				 * operation is valid.
				 */
    int bl_distance[TT_MAXTYPES];
} BloatData;

typedef struct squares_data
{
    int sq_border;
    int sq_size;
    int sq_sep;
    int sq_gridx;		/* Only used for "squares-grid" */
    int sq_gridy;		/* Only used for "squares-grid" */
} SquaresData;

typedef struct slots_data
{
    int sl_sborder;		/* short tile side */
    int sl_ssize;
    int sl_ssep;
    int sl_lborder;		/* long tile side */
    int sl_lsize;
    int sl_lsep;
    int sl_offset;
} SlotsData;

typedef struct cifop
{
    TileTypeBitMask co_paintMask;/* Zero or more paint layers to consider. */
    TileTypeBitMask co_cifMask;	 /* Zero or more other CIF layers. */
    int co_opcode;		/* Which geometric operation to use.  See
				 * below for the legal ones.
				 */
    int co_distance;		/* Grow or shrink distance (if needed). */
    ClientData co_client;	/* Pointer to a BloatData, SquaresData, or
				 * SlotsData structure, or NULL.
				 */
    struct cifop *co_next;	/* Next in list of operations to perform. */
} CIFOp;

/* The opcodes defined so far are:
 *
 * CIFOP_AND -		AND current results with the layers indicated by
 *			the masks.
 * CIFOP_ANDNOT -	Wherever there is material indicated by the masks,
 *			erase those areas from the current results.
 * CIFOP_OR -		OR current results with the layers indicated by
 *			the masks.
 * CIFOP_GROW -		Grow the current results uniformly by co_distance.
 * CIFOP_GROW_G -	Grow the current results to snap to the indicated grid.
 * CIFOP_SHRINK -	Shrink the current results uniformly by co_distance.
 * CIFOP_BLOAT -	Find layers in paintMask, then bloat selectively
 *			according to bl_distance, and OR the results into
 *			the current plane
 * CIFOP_SQUARES -	Generates a pattern of squares (used for making
 *			contact vias.  Each square is co_distance large,
 *			the squares are separated from each other by
 *			co_distance, and they are inside the edge of
 *			the material by at least co_distance.
 * CIFOP_SLOTS -	Generate a pattern of rectangles (used for making
 *			slots and slot vias).  Similar to squares except
 *			for different dimensions in short and long tile
 *			dimensions. "0" for the long size indicates that
 *			the slot should extend the length of the tile
 *			minus the long-side border length.
 * CIFOP_BLOATMAX -	Like CIFOP_BLOAT, except whole side of tile gets
 *			bloated by same amount, which is max bloat from
 *			anywhere along side.  Bloats can be negative.
 * CIFOP_BLOATMIN -	Same as CIFOP_BLOAT, except use min bloat from
 *			anywhere along side.
 * CIFOP_BLOATALL -	Added 3/21/05---bloat to encompass all connected
 *			material of the indicated type(s).
 * CIFOP_BBOX -		Added 4/2/05---create a single rectangle encompassing
 *			the cell bounding box.  This involves no magic type
 *			layers but may itself be acted upon with grow/shrink
 *			rules.
 */

#define CIFOP_AND	1
#define CIFOP_OR	2
#define CIFOP_GROW	3
#define CIFOP_GROW_G	4
#define CIFOP_SHRINK	5
#define CIFOP_BLOAT	6
#define CIFOP_SQUARES	7
#define CIFOP_SLOTS	8
#define CIFOP_BLOATMAX	9
#define CIFOP_BLOATMIN	10
#define CIFOP_BLOATALL	11
#define CIFOP_ANDNOT	12
#define CIFOP_SQUARES_G	13
#define CIFOP_BBOX	14

/* Added by Tim 10/21/2004 */
/* The following structure is used to pass information on how to draw
 * contact subcell arrays for a specific magic contact tile type.  For
 * the GDS write routine, the GDS file (FILE *) is passed as the client
 * data.
 */

typedef struct cifsquaresinfo
{
    SquaresData	*csi_squares;	/* Information on how to generate squares */
    TileType	 csi_type;	/* Magic contact tile type */
    ClientData	 csi_client;	/* Used to pass output file info.	*/
} CIFSquaresInfo;

/* The following data structure contains all the information about
 * a particular CIF layer.
 */

typedef struct
{
    char *cl_name;		/* Name of layer. */
    CIFOp *cl_ops;		/* List of operations.  If NULL, layer is
				 * determined entirely by cl_initial.
				 */
    int cl_growDist;		/* Largest distance material may move in
				 * this layer from its original Magic
				 * position, due to grows.  Expressed
				 * in CIF units.  If this layer uses temp
				 * layers, this distance must include grows
				 * from the temp layers.
				 */
    int cl_shrinkDist;		/* Same as above, except for shrinks. */
    int cl_flags;		/* Bunches of flags:  see below. */
    int cl_calmanum;		/* Number (0-63) of this layer for output as
				 * Calma (GDS-II stream format), or -1 if
				 * this layer should not be output.
				 */
    int cl_calmatype;		/* Data type (0-63) for Calma output, or -1
				 * if this layer should not be output.
				 */
    int min_width;		/* the minimum width rule in centi-microns
				 * for the layer. This is used by Grow Sliver
				 * to generate drc correct parent slivers	
				 */
#ifdef THREE_D
    int	   cl_renderStyle;	/* Style to render CIF layer with */
    float  cl_height;		/* (rendered) height of CIF layer above substrate */
    float  cl_thick;		/* (rendered) thickness of CIF layer */
#endif

} CIFLayer;

/* The CIFLayer flags are:
 *
 * CIF_TEMP:	Means that this is a temporary layer used to build
 *		up CIF information.  It isn't output in the CIF file.
 * CIF_BBOX_TOP:  Indicates that the bounding box rectangle should
 *		only be generated if the cell is a top-level cell.
 */

#define CIF_TEMP 	1
#define CIF_BBOX_TOP	2

/* The following data structure describes a complete set of CIF
 * layers.  The number of CIF layers (MAXCIFLAYERS) must not be
 * greater than the number of tile types (TT_MAXTYPES)!!
 */

#define MAXCIFLAYERS (TT_MAXTYPES - 1)

typedef struct cifkeep
{
    struct cifkeep	*cs_next;
    char		*cs_name;
} CIFKeep;

typedef struct cifstyle
{
    char cs_status;		/* Status:  Loaded, not loaded, or pending. */
    char *cs_name;		/* Name used for this kind of CIF. */
    int cs_nLayers;		/* Number of layers. */
    int cs_radius;		/* Radius of interaction for hierarchical
				 * processing (expressed in Magic units).
				 */
    int cs_stepSize;		/* If non-zero, user-specified step size
				 * for hierarchical processing (in Magic
				 * units).
				 */
    int cs_gridLimit;		/* The limit of grid scaling.  This limits
				 * the use of "scalegrid" to prevent Magic
				 * from generating geometry smaller than the
				 * process minimum grid.
				 */
    int cs_scaleFactor;		/* Number of CIF units per Magic unit.
 				 * CIF units are usually centimicrons, but
				 * see cs_expander below.
				 */
    int cs_reducer;		/* Reduction factor (used only to reduce
				 * number of zeroes in CIF files and make
				 * file more readable).  Default of 1.
				 * Unused for GDS input/output.
				 */
    int cs_expander;		/* cs_scaleFactor / cs_expander = scale in 
				 * centimicrons.  Default of 1.  Value 10
				 * means cs_scaleFactor is measured in
				 * nanometers (millimicrons)
                                 */

    TileTypeBitMask cs_yankLayers;
				/* For hierarchical processing, only these
				 * Magic types need to be yanked.
				 */
    TileTypeBitMask cs_hierLayers;
				/* For hierarchical processing, only these
				 * CIF layers need to be generated.
				 */
    int cs_labelLayer[TT_MAXTYPES];
				/* Each entry corresponds to one Magic layer,
				 * and gives index of CIF real layer to use
				 * for labels attached to this Magic layer.
				 * -1 means no known CIF layer for this Magic
				 * layer.
				 */
    CIFLayer *cs_layers[MAXCIFLAYERS];
				/* Describes how to generate each layer.*/
    int cs_flags;		/* bitmask of boolean-valued output options */

} CIFStyle;

/* values for cs_flags */
#define CWF_PERMISSIVE_LABELS	1
#define CWF_GROW_SLIVERS	2
#define CWF_ANGSTROMS 		4
#define CWF_GROW_EUCLIDEAN	8

/* procedures */

extern bool CIFNameToMask();
extern void CIFGenSubcells();
extern void CIFGenArrays();
extern void CIFGen();
extern void CIFClearPlanes();
extern Plane *CIFGenLayer();
extern void CIFInitCells();
extern int cifHierCopyFunc();
extern void CIFLoadStyle();

/* Shared variables and structures: */

extern Plane *CIFPlanes[];		/* Normal place to store CIF. */
extern CIFKeep *CIFStyleList;		/* List of all CIF styles. */
extern CIFStyle *CIFCurStyle;		/* Current style being used. */
extern CellUse *CIFComponentUse;	/* Flatten stuff in here if needed. */
extern CellDef *CIFComponentDef;	/* Corresponds to CIFComponentUse. */
extern CellUse *CIFDummyUse;		/* Used to dummy up a CellUse for a
					 * def.
					 */

/* Valid values of CIFWarningLevel (see cif.h) */

typedef enum {CIF_WARN_DEFAULT, CIF_WARN_NONE, CIF_WARN_ALIGN,
	CIF_WARN_LIMIT, CIF_WARN_REDIRECT, CIF_WARN_END} CIFWarningTypes;

/* Statistics counters: */

extern int CIFTileOps;
extern int CIFHierTileOps;
extern int CIFRects;
extern int CIFHierRects;

/* Tables used for painting and erasing CIF. */

extern PaintResultType CIFPaintTable[], CIFEraseTable[];

/* Procedures and variables for reporting errors. */

extern int CIFErrorLayer;
extern CellDef *CIFErrorDef;
extern void CIFError();

/* The following determines the tile type used to hold the CIF
 * information on its paint plane.
 */

#define CIF_SOLIDTYPE 1
extern TileTypeBitMask CIFSolidBits;

#endif /* _CIFINT_H */
